### PR TITLE
JIT: Support `__device__` function

### PR DESCRIPTION
**Pull request by asi1024**
Close https://github.com/cupy/cupy/issues/5421.

Usage:
```py
import numpy
import cupy
from cupy import testing
from cupyx import jit

@jit.rawkernel(device=True)
def getitem(x, tid):
    return x[tid]

@jit.rawkernel()
def elementwise_copy(x, y):
    tid = jit.threadIdx.x + jit.blockDim.x * jit.blockIdx.x
    y[tid] = getitem(x, tid)

x = testing.shaped_random((30,), dtype=numpy.int32, seed=0)
y = testing.shaped_random((30,), dtype=numpy.int32, seed=1)

elementwise_copy((5,), (6,), (x, y))
assert bool((x == y).all())

print(elementwise_copy.cached_code)
```

Output:
```cpp
__device__ int getitem_1(CArray<int, 1, true, true> x, unsigned int tid) {
  return x[tid];
}
extern "C" __global__ void elementwise_copy(CArray<int, 1, true, true> x, CArray<int, 1, true, true> y) {
  unsigned int tid;
  tid = (threadIdx.x + (blockDim.x * blockIdx.x));
  y[tid] = getitem_1(x, tid);
}
```

TODO:
- [x] Raise `Valuerror` for recursive function.
- [x] Write tests

### CODE REVIEW

File: tests/cupyx_tests/jit_tests/test_device_function.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    def test_device_function(self):
+        @jit.rawkernel()
+        def f(x, y, z):
+            tid = jit.threadIdx.x
+            z[tid] = g(x[tid], y[tid]) + x[tid] + y[tid]
+
+        def g(x, y):
+            x += 1
+            y += 1
+            return x + y
```
Comment (asi1024) [inline tests/cupyx_tests/jit_tests/test_device_function.py:21]: Note:
```cpp
__device__ int g_1(int x, int y) {
  x = (x + 1);
  y = (y + 1);
  return (x + y);
}
extern "C" __global__ void f(CArray<int, 1, true, true> x, CArray<int, 1, true, true> y, CArray<int, 1, true, true> z) {
  unsigned int tid;
  tid = threadIdx.x;
  z[tid] = ((g_1(x[tid], y[tid]) + x[tid]) + y[tid]);
}
```

File: tests/cupyx_tests/jit_tests/test_device_function.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        def f(x, y, z):
+            tid = jit.threadIdx.x
+            z[tid] = g(10)(x[tid], y[tid])
+            z[tid] += g(20)(x[tid], y[tid])
+            z[tid] += g(30)(x[tid], y[tid])
+
+        def g(const):
+            def f(x, y):
+                return x + y + const
+            return f
```
Comment (asi1024) [inline tests/cupyx_tests/jit_tests/test_device_function.py:40]: Note:
```cpp
__device__ int f_1(int x, int y) {
  return ((x + y) + 10);
}
__device__ int f_2(int x, int y) {
  return ((x + y) + 20);
}
__device__ int f_3(int x, int y) {
  return ((x + y) + 30);
}
extern "C" __global__ void f(CArray<int, 1, true, true> x, CArray<int, 1, true, true> y, CArray<int, 1, true, true> z) {
  unsigned int tid;
  tid = threadIdx.x;
  z[tid] = f_1(x[tid], y[tid]);
  z[tid] = (z[tid] + f_2(x[tid], y[tid]));
  z[tid] = (z[tid] + f_3(x[tid], y[tid]));
}
```

File: tests/cupyx_tests/jit_tests/test_device_function.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+        def g(step):
+            def f(x):
+                if step < 256:
+                    tid = jit.threadIdx.x
+                    if tid % (step * 2) == 0:
+                        x[tid] += x[tid + step]
+                        jit.syncthreads()
+                        g(step * 2)(x)
+            return f
```
Comment (asi1024) [inline tests/cupyx_tests/jit_tests/test_device_function.py:79]: Note:

<details>

```cpp
template <typename T0, typename T1>
__device__ unsigned int cupy_remainder_uint32(T0 in0, T1 in1) {
    return (in0 - _floor_divide(in0, in1) * in1) * (in1 != 0);
}


__device__ void f_9(CArray<int, 1, true, true> x) {
}
__device__ void f_8(CArray<int, 1, true, true> x) {
  unsigned int tid;
  tid = threadIdx.x;
  if ((cupy_remainder_uint32(tid, 256u) == 0u)) {
    x[tid] = (x[tid] + x[(tid + 128u)]);
    __syncthreads();
    f_9(x);
  }
  else {
  }
}
__device__ void f_7(CArray<int, 1, true, true> x) {
  unsigned int tid;
  tid = threadIdx.x;
  if ((cupy_remainder_uint32(tid, 128u) == 0u)) {
    x[tid] = (x[tid] + x[(tid + 64u)]);
    __syncthreads();
    f_8(x);
  }
  else {
  }
}
__device__ void f_6(CArray<int, 1, true, true> x) {
  unsigned int tid;
  tid = threadIdx.x;
  if ((cupy_remainder_uint32(tid, 64u) == 0u)) {
    x[tid] = (x[tid] + x[(tid + 32u)]);
    __syncthreads();
    f_7(x);
  }
  else {
  }
}
__device__ void f_5(CArray<int, 1, true, true> x) {
  unsigned int tid;
  tid = threadIdx.x;
  if ((cupy_remainder_uint32(tid, 32u) == 0u)) {
    x[tid] = (x[tid] + x[(tid + 16u)]);
    __syncthreads();
    f_6(x);
  }
  else {
  }
}
__device__ void f_4(CArray<int, 1, true, true> x) {
  unsigned int tid;
  tid = threadIdx.x;
  if ((cupy_remainder_uint32(tid, 16u) == 0u)) {
    x[tid] = (x[tid] + x[(tid + 8u)]);
    __syncthreads();
    f_5(x);
  }
  else {
  }
}
__device__ void f_3(CArray<int, 1, true, true> x) {
  unsigned int tid;
  tid = threadIdx.x;
  if ((cupy_remainder_uint32(tid, 8u) == 0u)) {
    x[tid] = (x[tid] + x[(tid + 4u)]);
    __syncthreads();
    f_4(x);
  }
  else {
  }
}
__device__ void f_2(CArray<int, 1, true, true> x) {
  unsigned int tid;
  tid = threadIdx.x;
  if ((cupy_remainder_uint32(tid, 4u) == 0u)) {
    x[tid] = (x[tid] + x[(tid + 2u)]);
    __syncthreads();
    f_3(x);
  }
  else {
  }
}
__device__ void f_1(CArray<int, 1, true, true> x) {
  unsigned int tid;
  tid = threadIdx.x;
  if ((cupy_remainder_uint32(tid, 2u) == 0u)) {
    x[tid] = (x[tid] + x[(tid + 1u)]);
    __syncthreads();
    f_2(x);
  }
  else {
  }
}
extern "C" __global__ void f(CArray<int, 1, true, true> x, CArray<int, 1, true, true> y) {
  unsigned int tid;
  tid = threadIdx.x;
  y[tid] = x[tid];
  __syncthreads();
  f_1(y);
}
```

**Comment by leofang** [comment]
FYI: recursive device function https://github.com/numba/numba/issues/7697

**Review by kmaehashi — APPROVED** [review]
LGTM!

**Pull request merged**